### ISSUE TITLE
Create SqlServer deployment guidance

### ISSUE THREAD

**Issue by tmasternak** [Particular/NServiceBus.SqlServer#180]
## Goal
Goal of this issue is to create documentation guiding `SqlServer` users in case of following requriements:
 * [ ] Requirement for better performance (probably overall throughput)
 * [ ] Requirement for high availability of the broker
 * [ ] Requirement for store-and-forward due to poor endpoint-to-broker connectivity

## Plan of Attack (draft):
- [ ] Spike on Sql Server 2016 Availability Groups
- [ ] Spike of Shovel-like agent #146 
- [ ] Document business scenarios that we want to use
- [ ] Create doco on using SQL Server availability groups

**Comment by tmasternak** [comment]
/cc: I'm not sure what label should we use? Maybe we need new one?

**Comment by MarcinHoppe** [comment]
@tmasternak I'd explore and document our recommendation for using In-Memory OLTP for table queues. Potentially huge performance gains.

**Comment by tmasternak** [comment]
@MarcinHoppe good point!

**Comment by MarcinHoppe** [comment]
@tmasternak I updated issue description.

**Comment by tmasternak** [comment]
@Particular/tf-v6-launch-plat-dev can you consider that for V6? /cc: @mauroservienti

**Comment by danielmarbach** [comment]
> Create SqlServer deployment guidance

sounds for me like a general purpose documentation like how to deploy SqlServer in general. I think this needs a more expressive title.

One thing that is not clear for me:

> If SQL Serve 2016 Availablity Groups turn out not to be sufficient. Refine plan and potentially look at

Sounds a bit off to me since if this item would be approved the scope of this item could suddenly change midway. How about keeping this one focused. When you find out that availability doesn't work, create a new item with a new PoA and get it approved/

**Comment by tmasternak** [comment]
> sounds for me like a general purpose documentation like how to deploy SqlServer in general. I think this needs a more expressive title

:+1:

**Comment by andreasohlund** [comment]
Having a migration path for v2 multi-instance users going to v3 seems like a must have for "core v6" since the spike might result in breaking changes needed in the transport.

@Particular/tf-v6-launch-plat-dev after talking to @danielmarbach and @tmasternak we decided to include this in the scope for v6


@tmasternak can you flesh out the description a bit more to explain why 2016 availability groups is the way to go?

**Comment by weralabaj** [comment]
@sfarmar also suggested that we should consider architectural aspects, e.g. before we said in the doco that this setup makes sense because endpoints are fully autonomous.

**Comment by johnsimons** [comment]
Should this include ServiceControl changes too ? So anyone that currently hacked SC to make it work, they need to revert ?

**Comment by johnsimons** [comment]
Does this include backwards compatibility, hot migrations, ... ?

**Comment by johnsimons** [comment]
Also, how does this new setup align with the ADSD guidelines ?

**Comment by mauroservienti** [comment]
> ADSD guidelines

which ADSD guidelines?

**Comment by johnsimons** [comment]
multiple databases

**Comment by johnsimons** [comment]
shared error and audit queue

**Comment by mikeminutillo** [comment]
For existing V2 users, shouldn't we have guidance around why to avoid Multi-Instance? We can talk about the difference between *the database* and *the broker*. In the case of SQL Transport it just so happens that the broker is also a database.

If a customer was architecting a system with Rabbit. They would be unlikely to have a separate rabbit broker per project and then try to connect them (wouldn't they?). The same logic applies here.

Put all your queues in a single database. Call it *the broker*. Make it highly available and scale it up as required using the built in SQL mechanisms. Put your business data elsewhere.

Is that advice correct? What impact does this advice have on outbox?

It the above it is correct then can we make this recommendation more clear in the documentation and should we reach out to existing SQL Transport customers to advice them of this? We have customers today that are getting to this state after months long projects and then coming to us for advice. The earlier we can advise them that this is the wrong course the better. Our only advice cannot be "wait for Core V6 and SQL V3 and then maybe multi-catalog will save you".

**Comment by tmasternak** [comment]
> Put all your queues in a single database. Call it the broker. Make it highly available and scale it up as required using the built in SQL mechanisms. Put your business data elsewhere.

I agree that we have to very explicit about distinguishing between SQL Server used as business data storage and transport mechanism.

> If a customer was architecting a system with Rabbit. They would be unlikely to have a separate rabbit broker per project and then try to connect them (wouldn't they?). The same logic applies here.

All though I would agree with the Rabbit case I can see users comparing SQL Server to MSMQ and treating it more like store-and-forward messaging technology.

**Comment by MarcinHoppe** [comment]
... and then there's also SQL Server backing NHibernate persistence, making it so much more fun overall. I think a document "How to architect an NServiceBus system with SQL Server" would be a very welcome addition to our doco.

**Comment by johnsimons** [comment]
@tmasternak we should have a meeting about this first thing Monday to get it sorted

**Comment by johnsimons** [comment]
Just to add more to what @mikeminutillo said, I see the schemas being an important part here, those are the equivalent of vhosts in rabbit or namespaces in azure, that is how user achieve "separation".
And the error and audit would probably leave in the default dbo.

**Comment by SzymonPobiega** [comment]
@mikeminutillo I would add "use outbox for exactly-once processing" in such a guidance. 

We acknowledge that SQL Server transport uses the broker model. Also, if, in future, we decide to properly add federation support to that transport, we will do it as transparently, as possible, via something like RabbitMQ's shovel. Still, the idea here would be not to have a broker node per endpoint, but per sub-system (e.g. distinct geography).

**Comment by SzymonPobiega** [comment]
I've created an initial draft [here](https://github.com/Particular/docs.particular.net/pull/1579)

**Comment by ramonsmits** [comment]
I would like to be part of the taskforce to verify the availability groups concept as a replacement for multi-instance support. I know that @mauroservienti was interested in creating such a lab environment. Who else? What is the priority to validate this concept?

**Comment by tmasternak** [comment]
Status
We close to being done with guidance [document ](https://github.com/Particular/docs.particular.net/pull/1579). We need to sync with @sfarmar and make sure we is fine with the content.

**Comment by sfarmar** [comment]
I've reviewed and added some comments, be happy to go on a call and discuss if needed

**Comment by MarcinHoppe** [comment]
It seems we can close this one? PR has been merged and other items got their separate issues.

Do we need a retrospective on this?

/cc @Particular/sqlserver-transport-maintainers

**Comment by johnsimons** [comment]
Awesome doco guys, well done, really good stuff there
